### ISSUE TITLE
Proposal: InformationFilter

### ISSUE THREAD

**Issue by koaning** [fairlearn/fairlearn#430]
I'm making this issue so that we can disucss how to move the `InformationFilter` from [scikit-fairness](https://scikit-fairness.netlify.app/api/preprocessing.html#skfair.preprocessing.InformationFilter) to this project. While moving it to this project I could also add a [guide](https://scikit-fairness.netlify.app/fairness_boston_housing.html#How-it-Works) to the docs on the boston housing dataset (the same we have on scikit-lego side). The details of the approach can also be seen on youtube [here](https://youtu.be/Z8MEFI7ZJlA?t=948). 

The goal of this issue is to first gather feedback on the implementation and, assuming people agree this tool is useful, to discuss the steps to make it compatible for this project. Once agreed on the path forward I'll open up a PR.

**Comment by romanlutz** [comment]
Thanks for sharing the video!

Adding @MiroDudik and @hannawallach as well. I think it's interesting, but we need to be very careful in the description / documentation. De-biasing data is usually a tricky topic since there are so many different kinds, and they can affect each other etc., so the way we've been looking at it so far is described in the [documentation](https://fairlearn.github.io/user_guide/fairness_in_machine_learning.html) (quoted below). That's much more focused on "harms", that is looking at the outcome or result.
That said, applying this could reduce the unfairness in the outcome at the end of the pipeline. We've discussed adding preprocessing techniques a few times, but never quite got down to it. I think it would be great to add it. 


> AI systems can behave unfairly for a variety of reasons. Sometimes it is because of societal biases reflected in the training data and in the decisions made during the development and deployment of these systems. In other cases, AI systems behave unfairly not because of societal biases, but because of characteristics of the data (e.g., too few data points about some group of people) or characteristics of the systems themselves. It can be hard to distinguish between these reasons, especially since they are not mutually exclusive and often exacerbate one another. Therefore, we define whether an AI system is behaving unfairly in terms of its impact on people — i.e., in terms of harms — and not in terms of specific causes, such as societal biases, or in terms of intent, such as prejudice.
Usage of the word bias. Since we define fairness in terms of harms rather than specific causes (such as societal biases), we avoid the usage of the words bias or debiasing in describing the functionality of Fairlearn.

**Comment by koaning** [comment]
It's certainly valid to point at the main weakness of the approach: technically it only applies a linear filter. Any non-linear effects may still be in there. It can be mitigated by adding a  `PolynomialFeatures` preprocessing step before filtering, but it remains a caveat. 

It's also valid to point out that you're merely filtering based on what you see in your dataset, which is most likely a biased dataset to even start with.

**Comment by romanlutz** [comment]
Exactly! I think if we do a good job in accurately documenting it then that's a good step forward. 
I'm sure we can find an example where the addition of preprocessing helps overall, or perhaps preprocessing + mitigation (e.g. postprocessing) helps in terms of the observed unfairness in the outcome. That actually fits rather nicely into the "harms" perspective.

**Comment by koaning** [comment]
Are there any objections to adding the `InformationFilter` as an approach? If not, then I can dive into the details of how `fairlearn` handles the sensitive attribute such that my proposal will be compatible code-wise.

**Comment by kevinrobinson** [comment]
I'd love to understand an example scenario (or help collaborate on fleshing out one)!  I've read https://scikit-fairness.netlify.app/fairness_boston_housing.html, watched the talk, and seen other great writing like [this one](https://koaning.io/posts/cost-of-success/), thanks for sharing! 👍 
![image](https://user-images.githubusercontent.com/1056957/82670296-e1325680-9c0a-11ea-8b6f-e48db77ef529.png)

We've chatted a bit about the Boston housing dataset in and of itself, but with regards to `InformationFilter` specifically I can't really follow why you would apply it in that scenario.  For example, if someone wants to predict median home value in a geographic area, why would they want to enforce that those predictions about real estate price are not correlated with incomes for that geographic area?  I'm also trying to understand how using `InformationFilter` to remove information about racial demographics in a geographic area connects with assessing or mitigating real-world harms related to housing.  Finding another scenario might just be all we need :)

@koaning it sounds like there might be other real world contexts where you have used this approach?  Alternately, picking another scenario that uses data points on individual people may work better for illustrating the value.  I'm happy to help brainstorm.

**Comment by koaning** [comment]
I wouldn’t mind picking another example. Part of the reason why we made the dataset fair against income there is because the effect is more pronounced. I would argue one needs to be careful with incomelevels for house prediction though. It can cause a self fulfilling prophecy.

Hypotetical Example

If one would use this dataset to automate house price prediction. If a low income neighborhood is getting gentrifed then this model may cause low income households to get less money for their house because they’re stamped as low-income.

**Comment by MiroDudik** [comment]
I saw the video, but I was wondering if there's a paper or tech report about this--not a requirement, but it might shorten the discussion.

Your `InformationFilter` reminds me of [Impartial predictive modeling: Ensuring fairness in arbitrary models](https://arxiv.org/pdf/1608.00528.pdf). Their approach in Definition 4 is equivalent to applying `InformationFilter` at training time, but then using the original features at the test time (without the sensitive features partialed out).

But I think that the way you [define fairness](https://scikit-fairness.netlify.app/fairness_boston_housing.html#Measuring-fairness-for-Regression) in your example is more like [Controlling Attribute Effect in Linear Regression](http://web.lums.edu.pk/~akarim/pub/controlling_attribute_effect_icdm2013.pdf). I think that you'd like to ensure, at least for (generalized) linear models, that the model prediction is uncorrelated with sensitive features, right? But if that's what you want then I think that you need to subtract feature means before doing Gram-Schmidt.

Did I get this right or is there something else that you had in mind to achieve with `InformationFilter`?

**Comment by koaning** [comment]
This approach was partially inspired from [this pydata talk](https://www.youtube.com/watch?v=AGgCqpouKSs) but also by thinking about linear algebra operations in general. After implementing it I didn't check the theory when I had evidence it worked in practice. It may very well be that there's a paper with this idea in mind and I wouldn't mind linking to related works, but this implementation is not based on a paper. I think it was just a whiteboard session I had with @Mbrouns.

You could certainly standardise with a `StandardScaler` or a `QuantileScaler` before passing data to the `InformationFilter` (this is often a good idea even without thinking about fairness) but even without it we've been able to see that it indeed removes correlation with sensitive attributes.

I would argue here that the way you prefer to quantify fairness is indepedant of the methods you apply. It may be that there's tools that specialize towards a certain fairness metric but the user can varify emperically if their preferred fairness measure gives good scores when the information filter (or any tool for fairness) is used.

**Comment by MiroDudik** [comment]
I completely agree that the mitigation technique can be chosen independently from the choice of the metric and the usefulness is driven by the evaluation criteria. I was mostly curious why you do things the way you do them in `InformationFilter`.

I generally like the idea and I think it would be a nice addition, but I think that the centering step should be always done.

One example to think about is:
* feature1 = [0, 0, 1, 1]
* feature2 = [1, 1, 2, 2]

If feature1 is sensitive, then using just Gram-Schimdt without recentering will give you
* feature2' = [1, 1, 0, 0]

That seems undesirable. If you center first, then you'll get feature2'= [0, 0, 0, 0].

**Comment by adrinjalali** [comment]
> I generally like the idea and I think it would be a nice addition, but I think that the centering step should be always done.

That's fair, but as @koaning mentioned, s scaler can be applied in the pipeline before this step. We should just document that the data is expected to be centered. You could also raise a warning if the input data deviates too much from what `InformationFilter` expects. It's usually much cleaner and nicer if the preprocessing needed for a step is not done in the step itself.

**Comment by MiroDudik** [comment]
In principle I am okay with raising a warning, I just want to make sure that the absence of centering is a feature not a bug :-). That's why I wanted more clarity about the intended behavior--the intended behavior should be described as a relationship between input and output rather than as a procedure.

If I understand this correctly, fairness here is formalized as statistical independence--the initial goal is to obtain transformed features that are statistically independent from the sensitive features, while keeping as much information as possible otherwise. If such a thing was possible, then any classifier / regressor based on the transformed features would also be statistically independent of the sensitive features.

Instead of going after this in full generality, one could simply require that the transformed features are uncorrelated with the sensitive features. In symbols, if `x` is the original feature vector and `s` is a vector of sensitive features, the goal is to derive a transformed feature vector `z` such that:

* `E[z s^T] = E[z] E[s]^T`
   in words: every component of `z` is uncorrelated with every component of `s`

That requires centering + Gram-Schmidt. And without centering you only achieve

* `E[z s^T] = 0`
   I'm not sure how this achieves independence in the absence of centering

I cannot easily think of any use case where one would use Gram-Schmidt without centering.

The pattern of seemingly combining two steps into one already exists in `sklearn`. In fact `StandardScaler` is doing two things--it is centering and re-scaling by default, and you can disable either of the two aspects if you like. I'd much rather follow a similar pattern here, where we center by default, and you can disable it. Do we need a call to figure this out?

**Comment by koaning** [comment]
I'd prefer to run some benchmarks just to confirm the confirm the effect of centering, but assuming it's an improvement I'm fine with centering in the method and having a flag for it that you can manually turn off if you wish. 

For sparse data though ... it may also make sense to make a seperate filter for that. I haven't explored sparse data at all here but I find it plausible that you may want to handle that use-case slightly differently.

**Comment by koaning** [comment]
I just ran some benchmark locally, I think centering should be in. I'll allow a user to turn it off but the effect, albeit subtle, is clear. Without it you'd get a filter that will grab the high correlation but you'd allow the perfect correlation to pass, which is wrong. 

@MiroDudik good observation. I still think that we might want to make a seperate component for sparse but I'll work on a first PR for this component.

**Comment by MiroDudik** [comment]
Thanks for trying this out! Did you have any thoughts on the modified Gram-Schmidt? It implements the idea "remove correlation without recentering". I initially thought of it in the context of sparsity, but I think it makes sense as the main approach too. In that case, I don't think we would need to center--and we would still be removing all the correlation--so it would be a conceptually simpler transformation.

**Comment by koaning** [comment]
That certainly something that we can still discuss, but I figured making an initial PR to also get the conversation started on testing was a good step to take at this phase. 

Suppose we have these two approaches, how might you benchmark them?

**Comment by koaning** [comment]
@MiroDudik I'm looking at your formula that you passed earlier. 

![](https://camo.githubusercontent.com/dd942f464e19dcca69fd5eabff3c60654af18e7b/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f6769662e6c617465783f7a253230253344253230782532302d2532302535436672616325374225354366726163316e2532307825354554253230732532302d253230253543626172253742782537442535436261722537427325374425374425374225354366726163316e2532307325354554253230732532302d253230253543626172253742732537442535436261722537427325374425374425354363646f7425323073)

How might you extend this in the case that you have more than one sensitive attribute? Let's say you have zipcode and gender?

**Comment by MiroDudik** [comment]
Let me write this up more formally.

Agreed re. benchmarking--the basic idea would be to look at statistical independence between Z and S vs. how much information from X is preserved in Z. We could do it in the context of a specific task or tasks--i.e., for some specific Y we wish to predict (as you do in your worked out example) or in terms of how well we can predict X from Z.

**Comment by koaning** [comment]
Whoa ... Extensive 👍 

I think the centered approach that is in the information filter right now is eventually equivalent but your approach has more vectorisation that can happen and because there's less operations it should also be numerically more stable. 

I just wrote a quick implementation, it's less code and substantially faster. 

<img width="964" alt="image" src="https://user-images.githubusercontent.com/1019791/84436379-a46ae580-ac33-11ea-8bd8-8bc446fd04af.png">

There's subtle differences in the numbers. I'm interested in understanding what is causing that but for now I think I can go ahead an rewrite the implementation. I'll add unit tests this weekend.

**Comment by koaning** [comment]
I've been running some benchmarks on the `load_boston` dataset and I fear we have to re-evaluate the method. I just found a huge difference between the original `InformationFilter` that uses Gram-Smidt and the `CorrelationRemover` that uses regression internally. 

Each gif represents the model output. I am taking the boston dataset and applying a linear regression after filtering away information. I've made two gifs that each display three charts. 

1. The 1st chart shows predictions on the x-axis and true values on the y-axis. 
2. The 2nd chart shows predicted values split on a sensitive attribute `B`.
3. The 3rd chart shows predicted values split on a sensitive attribute `LSTAT`.

Here's a gif that displays the effect of alpha on the `Correlation Filter` 

![boston_corr](https://user-images.githubusercontent.com/1019791/85955726-f2e0e980-b980-11ea-9e56-a546565c2889.gif)

Here's a gif that displays the effect of alpha on the `InformationFilter`.

![boston_filter](https://user-images.githubusercontent.com/1019791/85955728-f5434380-b980-11ea-8fcf-c10337c29b52.gif)

Looking at these gifs I noticed that both approaches made a changer but that the information filter was much more "aggressive". So I decided to make another plot.



![image](https://user-images.githubusercontent.com/1019791/85955755-2c195980-b981-11ea-8003-bbabc657b56d.png)

**Comment by koaning** [comment]
Here's the notebook; https://gist.github.com/koaning/1a606e07c9b27d1889ada4185959d87a

**Comment by koaning** [comment]
Either way, I seem to have learned a lesson for the hyperparameter ... it should not be constrained to (0, 1)

**Comment by koaning** [comment]
I might have a proposal. Shall we make two transformers. One `GramSmidthFilter` and a `CorrelationRemover`?

**Comment by romanlutz** [comment]
I'm open to that, but we really need this kind of analysis in the user guide I think. Users need to understand when to use which one.

**Comment by koaning** [comment]
@MiroDudik any comments? I think my next step is to write up a proper benchmark. @romanlutz I can run this on boston with linear regression, knn and some other models. got a suggestion for another dataset?

**Comment by MiroDudik** [comment]
Interesting! I'm not sure I understand. Are you standardizing the features before transforming with `InformationFilter`? In that case the two approaches should be mathematically equivalent (but numerically they might behave differently). Also, if I understand your benchmark plots correctly, you are plotting the differences between mean predictions between the two groups? Can you also plot the covariances between predictions and sensitive features? Before deciding next steps, I'll definitely want to check the notebook to have a better sense what's going on.

**Comment by koaning** [comment]
@MiroDudik agreed. I'll see if I can make a more clear/elaborate notebook and share it here.

**Comment by MiroDudik** [comment]
I see--if you don't center as the first step then they are not equivalent. I'll wait for the notebook!

**Comment by koaning** [comment]
The notebook can be found [here](https://gist.github.com/koaning/bc628caf8f7d905891e8886e7c574fcf).

This chart shows the main conclusion; 

![image](https://user-images.githubusercontent.com/1019791/86493256-218c0500-bd71-11ea-9473-71852db4dad6.png)

I've ignored the accuracy metrics for now because that is not the focus.

**Comment by romanlutz** [comment]
@koaning you could use the adult dataset as well. I have a PR out right now to add it to the datasets module, but the relevant code is just that one line if you don’t want to wait.

**Comment by koaning** [comment]
This seems to be the 1st chart for the Adult dataset. Different results, but also because the dataset is very different. This is a model with a lot of dummy variables. 

![image](https://user-images.githubusercontent.com/1019791/86634533-43d08d80-bfd2-11ea-8ddf-732499c55299.png)

**Comment by koaning** [comment]
Updated the notebook here with code for the ADULT dataset https://gist.github.com/koaning/bc628caf8f7d905891e8886e7c574fcf.

**Comment by koaning** [comment]
@romanlutz @MiroDudik ping :)

**Comment by MiroDudik** [comment]
@koaning : apologies for the inordinate delay--are you still interested in seeing this through? I can commit a few hours each week to make sure that we make steady progress.

**Comment by MiroDudik** [comment]
There were two bugs in the notebook. One of them was in the implementation of `CorrelationRemover.fit()`:
* _original_: `self.sensitive_mean_ = X_sensitive.mean()`
  _correct_: `self.sensitive_mean_ = X_sensitive.mean(axis=0)`

The second was in `run_stat()` implementation on **adult**:
* _original_:
```python
    pipe = Pipeline([
        ('filter', filt([0, 1], alpha=alpha)),
        ('mod', alg)
    ])
```
* _correct_:
```python
    pipe = Pipeline([
        ('filter', filt([1, 2], alpha=alpha)),
        ('mod', alg)
    ])
```

With those two bugs fixed, some of the paradoxes that we saw previously disappear. To generate the plots below, I was running [this notebook](https://gist.github.com/MiroDudik/0eb70f05631fb0c240cf3aa29c7d1576).

Besides the disparity metrics that you measured previously, I'm also measuring a rescaled version of covariance (scaled to be comparable to the difference stats). Similarly to the difference stats, covariance=0 here corresponds to greater fairness:

### Boston dataset
![image](https://user-images.githubusercontent.com/20565444/94611075-38a69a00-026f-11eb-9f72-1b2f0630d55e.png)
### Adult dataset
![image](https://user-images.githubusercontent.com/20565444/94611909-6213f580-0270-11eb-95fc-0fbb75fa0143.png)

Linear models (linear regression and logistic regression) behave as expected: `CorrelationRemover` leads to the greatest fairness with alpha close to 1. On the other hand, behavior of `InformationRemover` is less consistent across the two linear models and it does not achieve the same low values of unfairness.

I'm not quite sure what to say about non-linear models except that neither of the two mitigation techniques does particularly well, except for SVR, which appears more similar to a linear model than to the k-nearest neighbors model (but that could be driven by the default choice of hyperparameters).

**Comment by koaning** [comment]
Ah yeah. Good catch there. Covariance is also a nice touch. 

I've confirmed these results locally and they convince me to focus on the `CorrelationRemover` part. It's faster, numerically more stable and currently I can't come up with edge cases where the `InformationFilter` approach will be better. If we agree that this `CorrelationRemover` is an appropriate addition I wouldn't mind preparing a PR. 

That said, some folks have been playing with the scikit-lego implementation and it does demonstrate some [other drawbacks](https://twitter.com/davidmasip2/status/1290644405880324096). If the orignal dataset is already balanced, it seems like this method might make it "unbalanced" and it seems to "hickup" when there's integers around (as opposed to floats floats/categoricals). The method still seems valid, but the docs should reflect that it is to be applied in combination with linear systems and that feedback is perhaps appreciated. 

@MiroDudik this is green for a PR?

**Comment by MiroDudik** [comment]
Yes--please go ahead and start `CorrelationRemover` as a PR. In the documentation, we should mention that this is intended as a preprocessing step for (generalized) linear models, and that "no correlation" does not mean "independence". If you really want to stress the point, we could even rename `CorrelationRemover` to `LinearDependenceRemover` (but that's really a mouthful).

I really like the observation how integer-valued features are impacted--a very nice illustration that the lack of correlation does not mean lack of dependence and that dependence and correlation can go in the opposite directions.

**Comment by romanlutz** [comment]
Closing this since the PR went in.
